File: mifosng-android/src/main/java/com/mifos/mifosxdroid/LoginActivity.java
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */
package com.mifos.mifosxdroid;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import com.mifos.mifosxdroid.base.BaseActivity;
import com.mifos.mifosxdroid.databinding.LoginBinding;
import com.mifos.mifosxdroid.online.DashboardFragmentActivity;
import com.mifos.objects.User;
import com.mifos.services.API;
import com.mifos.utils.Constants;
import com.mifos.utils.MifosApplication;
import com.mifos.utils.ValidationUtils;

import org.apache.http.HttpStatus;

import java.net.MalformedURLException;
import java.net.URISyntaxException;
import java.net.URL;

import javax.net.ssl.SSLHandshakeException;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

import static android.view.View.GONE;
import static android.view.View.OnClickListener;
import static android.view.View.VISIBLE;
import static com.mifos.utils.Constants.API_PATH;
import static com.mifos.utils.Constants.PROTOCOL_HTTPS;

/**
 * Created by ishankhanna on 08/02/14.
 */
public class LoginActivity extends BaseActivity implements Callback<User> {
    private Integer port = null;
    private String username;
    private String instanceURL;
    private String password;

    private API api;
    private LoginBinding binding;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_login);

        String previouslyEnteredUrl = MifosApplication.spManager.getInstanceUrl();
        String previouslyEnteredPort = MifosApplication.spManager.getInstancePort();

        if (!previouslyEnteredUrl.startsWith("http"))
            previouslyEnteredUrl = Constants.PROTOCOL_HTTPS + previouslyEnteredUrl;

        binding.etInstanceURL.setText(previouslyEnteredUrl);

        if (!previouslyEnteredPort.equals("80"))
            binding.etInstancePort.setText(previouslyEnteredPort);

        updateMyInstanceUrl();

        binding.btConnectionSettings.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                if (binding.llConnectionSettings.getVisibility() == VISIBLE)
                    binding.llConnectionSettings.setVisibility(GONE);
                else
                    binding.llConnectionSettings.setVisibility(VISIBLE);
            }
        });

        binding.btLogin.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                binding.passwordWrapper.setErrorEnabled(false);
                binding.userWrapper.setErrorEnabled(false);
                binding.urlWrapper.setErrorEnabled(false);

                username = binding.etUsername.getEditableText().toString();
                if (username.length() < 5) {
                    binding.userWrapper.setError("Username is not valid");
                    return;
                }
                password = binding.etPassword.getEditableText().toString();
                if (password.length() < 6) {
                    binding.passwordWrapper.setError("Password is not valid");
                    return;
                }

                instanceURL = binding.etInstanceURL.getEditableText().toString();
                if (!TextUtils.isEmpty(binding.etInstancePort.getEditableText()))
                    port = Integer.parseInt(binding.etInstancePort.getEditableText().toString());


                String domain = ValidationUtils.sanitizeDomainNameInput(instanceURL);

                if (!ValidationUtils.validateURL(domain)) {
                    binding.urlWrapper.setError("Domain name is not valid");
                    return;
                }

                try {
                    String constructedURL = constructInstanceUrl(domain, port);
                    URL url = new URL(constructedURL);
                    instanceURL = url.toURI().toString();
                } catch (MalformedURLException e) {
                    binding.urlWrapper.setError("Url is not valid");
                    return;
                } catch (URISyntaxException uriException) {
                    binding.urlWrapper.setError("Url is not valid");
                    return;
                }

                login(false);
            }
        });
        binding.etInstanceURL.addTextChangedListener(urlWatcher);
        binding.etInstancePort.addTextChangedListener(urlWatcher);
    }

    private TextWatcher urlWatcher = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @Override
        public void afterTextChanged(Editable s) {
            updateMyInstanceUrl();
        }
    };

    private void login(boolean shouldByPassSSLSecurity) {
        showProgress("Logging in");

        api = new API(instanceURL, binding.etTenantIdentifier.getEditableText().toString().trim(), shouldByPassSSLSecurity);
        api.userAuthService.authenticate(username, password, this);
    }

    private void updateMyInstanceUrl() {
        String textUnderConstruction;

        if (!binding.etInstancePort.getEditableText().toString().isEmpty())
            port = Integer.valueOf(binding.etInstancePort.getEditableText().toString().trim());

        instanceURL = binding.etInstanceURL.getEditableText().toString();
        String domain = ValidationUtils.sanitizeDomainNameInput(instanceURL);
        textUnderConstruction = constructInstanceUrl(domain, port);

        binding.tvConstructedInstanceUrl.setText(textUnderConstruction);
    }

    public String constructInstanceUrl(String validDomain, Integer port) {
        return PROTOCOL_HTTPS + validDomain + (port != null ? (":" + port) : "") + API_PATH;
    }

    @Override
    public void success(User user, Response response) {
        MifosApplication.api = api;
        hideProgress();

        Snackbar.make(binding.getRoot(), getString(R.string.toast_welcome) + " " + user.getUsername(), Snackbar.LENGTH_SHORT).show();

        MifosApplication.spManager.setInstanceUrl(ValidationUtils.sanitizeDomainNameInput(binding.etInstanceURL.getEditableText().toString()));
        MifosApplication.spManager.setInstanceDomain(ValidationUtils.sanitizeDomainNameInput(binding.etInstanceURL.getEditableText().toString()));

        if (port != null)
            MifosApplication.spManager.setInstancePort(port.toString());

        String tenant = binding.etTenantIdentifier.getEditableText().toString();
        if (TextUtils.isEmpty(tenant))
            tenant = "default";

        MifosApplication.spManager.setTenantId(tenant);
        MifosApplication.spManager.setAuthToken("Basic " + user.getBase64EncodedAuthenticationKey());
        startActivity(new Intent(LoginActivity.this, DashboardFragmentActivity.class));
        finish();
    }

    @Override
    public void failure(RetrofitError retrofitError) {
        try {
            hideProgress();
            if (retrofitError.getCause() instanceof SSLHandshakeException) {
                promptUserToByPassTheSSLHandshake();
            } else if (retrofitError.getResponse().getStatus() == HttpStatus.SC_UNAUTHORIZED)
                Snackbar.make(binding.getRoot(), R.string.error_login_failed, Snackbar.LENGTH_LONG).show();
        } catch (NullPointerException e) {
            Snackbar.make(binding.getRoot(), R.string.error_unknown, Snackbar.LENGTH_LONG).show();
        }
    }

    /**
     * This method should show a dialog box and ask the user
     * if he wants to use and unsafe connection. If he agrees
     * we must update our rest adapter to use an unsafe OkHttpClient
     * that trusts any damn thing.
     */
    private void promptUserToByPassTheSSLHandshake() {
        AlertDialog alertDialog = new AlertDialog.Builder(this)
                .setTitle("SSL Certificate Problem")
                .setMessage("There is a problem with your SSLCertificate, would you like to continue? This connection would be unsafe.")
                .setIcon(android.R.drawable.stat_sys_warning)
                .setPositiveButton("Continue", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        login(true);
                    }
                })
                .setNegativeButton("Exit", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        finish();
                    }
                })
                .create();
        alertDialog.show();

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.offline_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.offline:
                startActivity(new Intent(this, OfflineCenterInputActivity.class));
                break;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: mifosng-android/src/main/java/com/mifos/ui/activities/LoginActivity.kt
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */
package com.mifos.ui.activities

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.InputType
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModelProvider
import com.mifos.api.BaseApiManager
import com.mifos.mifosxdroid.HomeActivity
import com.mifos.mifosxdroid.R
import com.mifos.mifosxdroid.core.MifosBaseActivity
import com.mifos.mifosxdroid.core.util.Toaster
import com.mifos.mifosxdroid.databinding.ActivityLoginBinding
import com.mifos.mifosxdroid.passcode.PassCodeActivity
import com.mifos.states.LoginUiState
import com.mifos.utils.Constants
import com.mifos.utils.Network
import com.mifos.utils.PrefManager
import com.mifos.utils.PrefManager.savePostAuthenticationResponse
import com.mifos.utils.PrefManager.saveToken
import com.mifos.utils.ValidationUtil
import com.mifos.viewmodels.LoginViewModel
import dagger.hilt.android.AndroidEntryPoint
import org.apache.fineract.client.models.PostAuthenticationResponse
import javax.inject.Inject

/**
 * Created by ishankhanna on 08/02/14.
 */
@AndroidEntryPoint
class LoginActivity : MifosBaseActivity(){

    private lateinit var binding: ActivityLoginBinding

    private lateinit var viewModel: LoginViewModel

    @Inject
    lateinit var baseApiManager: org.mifos.core.apimanager.BaseApiManager

    private lateinit var username: String
    private lateinit var instanceURL: String
    private lateinit var password: String
    private lateinit var domain: String
    private var isValidUrl = false
    private val urlWatcher: TextWatcher = object : TextWatcher {
        override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
        override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
        override fun afterTextChanged(editable: Editable) {
            val port = if (binding.etInstancePort.editableText.toString()
                    .isEmpty()
            ) null else Integer.valueOf(binding.etInstancePort.editableText.toString())
            instanceURL = ValidationUtil.getInstanceUrl(binding.etInstanceURL.text.toString(), port)
            isValidUrl = ValidationUtil.isValidUrl(instanceURL)
            binding.tvConstructedInstanceUrl.text = instanceURL
            domain = binding.etInstanceURL.editableText.toString()
            if (domain.isEmpty() || domain.contains(" ")) {
                isValidUrl = false
            }
            binding.tvConstructedInstanceUrl.setTextColor(
                if (isValidUrl) ContextCompat.getColor(
                    applicationContext, R.color.green_light
                ) else ContextCompat.getColor(
                    applicationContext, R.color.red_light
                )
            )
        }
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        title = null
        setContentView(binding.root)
        binding.etInstancePort.inputType = InputType.TYPE_CLASS_NUMBER
        if (PrefManager.getPort() != "80") binding.etInstancePort.setText(PrefManager.getPort())
        binding.etInstanceURL.setText(PrefManager.getInstanceDomain())
        binding.etInstanceURL.addTextChangedListener(urlWatcher)
        binding.etInstancePort.addTextChangedListener(urlWatcher)
        urlWatcher.afterTextChanged(Editable.Factory.getInstance().newEditable(""))

        viewModel = ViewModelProvider(this)[LoginViewModel::class.java]

        binding.btLogin.setOnClickListener {
            hideKeyboard(binding.btLogin)
            login()
        }

        binding.etPassword.setOnEditorActionListener { _, actionId, keyEvent ->
            if (actionId == EditorInfo.IME_ACTION_DONE || (keyEvent != null && keyEvent.keyCode == KeyEvent.KEYCODE_ENTER && keyEvent.action == KeyEvent.ACTION_DOWN)) {
                login()
                return@setOnEditorActionListener true
            }
            return@setOnEditorActionListener false
        }

        viewModel.loginUiState.observe(this){
            when(it) {
                is LoginUiState.ShowProgress -> showProgressbar(it.state)
                is LoginUiState.ShowLoginSuccessful -> {
                    hideProgress()
                    onLoginSuccessful(it.user)
                }
                is LoginUiState.ShowError -> {
                    hideProgress()
                    onLoginError(it.message)
                }
            }
        }
    }

    private fun validateUserInputs(): Boolean {
        domain = binding.etInstanceURL.editableText.toString()
        if (domain.isEmpty() || domain.contains(" ")) {
            showToastMessage(getString(R.string.error_invalid_url))
            return false
        }
        if (!isValidUrl) {
            showToastMessage(getString(R.string.error_invalid_connection))
            return false
        }
        username = binding.etUsername.editableText.toString()
        password = binding.etPassword.editableText.toString()
        if (username.isEmpty() || password.isEmpty()) {
            showToastMessage(getString(R.string.error_enter_credentials))
            return false
        } else {
            if (username.length < 5) {
                showToastMessage(getString(R.string.error_username_length))
                return false
            }
            if (password.length < 6) {
                showToastMessage(getString(R.string.error_password_length))
                return false
            }
        }
        return true
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_login, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle item selection
        return when (item.itemId) {
            R.id.mItem_connection_settings -> {
                binding.llConnectionSettings.visibility =
                    if (binding.llConnectionSettings.visibility == View.VISIBLE) View.GONE else View.VISIBLE
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun showToastMessage(message: String) {
        Toaster.show(findViewById(android.R.id.content), message, Toaster.LONG)
    }

    private fun onLoginSuccessful(user: PostAuthenticationResponse) {

        PrefManager.usernamePassword = Pair(username,password)
        // Saving userID
        PrefManager.setUserId(user.userId!!.toInt())
        // Saving user's token
        saveToken("Basic " + user.base64EncodedAuthenticationKey)
        // Saving user
        savePostAuthenticationResponse(user)
        Toast.makeText(
            this, getString(R.string.toast_welcome) + " " + user.username, Toast.LENGTH_SHORT
        ).show()
        if (PrefManager.getPassCodeStatus()) {
            startActivity(Intent(this, HomeActivity::class.java))
        } else {
            val intent = Intent(this, PassCodeActivity::class.java)
            intent.putExtra(Constants.INTIAL_LOGIN, true)
            startActivity(intent)
        }
        finish()
    }

    private fun onLoginError(errorMessage: String) {
        showToastMessage(errorMessage)
    }

    private fun showProgressbar(show: Boolean) {
        if (show) {
            showProgress(getString(R.string.logging_in))
        } else {
            hideProgress()
        }
    }


    private fun login() {
        if (!validateUserInputs()) {
            return
        }
        // Saving tenant
        PrefManager.setTenant(binding.etTenantIdentifier.editableText.toString())
        // Saving InstanceURL for next usages
        PrefManager.setInstanceUrl(instanceURL)
        // Saving domain name
        PrefManager.setInstanceDomain(binding.etInstanceURL.editableText.toString())
        // Saving port
        PrefManager.setPort(binding.etInstancePort.editableText.toString())
        // Updating Services
        BaseApiManager.createService()
        baseApiManager.createService(username,password,PrefManager.getInstanceUrl(),PrefManager.getTenant(),true)
        if (Network.isOnline(this)) {
            viewModel.login(username, password)
        } else {
            showToastMessage(getString(R.string.error_not_connected_internet))
        }
    }

}